consensus/snapshot/local_snapshots/local_snapshots_manager.c

```diff
@@ -0,0 +1,175 @@
+
+  lock_handle_init(&lock_cond);
+  lock_handle_lock(&lock_cond);
+
+  tangle_t tangle;
+  connection_config_t db_conf = {.db_path = lsm->conf->db_path};
+  if ((err = iota_tangle_init(&tangle, &db_conf))) {
+    log_critical(logger_id, "Failed in initializing db\n");
+    return NULL;
+  }
+
+  bool skip_check = false;
+
+  while (lsm->running) {
+    if (skip_check || iota_local_snapshots_manager_should_take_snapshot(lsm, &tangle)) {
+      start_timestamp = current_timestamp_ms();
+      err = iota_snapshots_service_take_snapshot(lsm->snapshots_service, &tangle);
+      if (err == RC_OK) {
+        exponential_delay_factor = 1;
+        end_timestamp = current_timestamp_ms();
+        log_info(logger_id, "Local snapshot completed successfully in %" PRIu64 " ms\n",
+                 end_timestamp - start_timestamp);
+        if (iota_tangle_transaction_count(&tangle, &lsm->last_snapshot_transactions_count) != RC_OK) {
+          log_critical(logger_id, "Failed in querying db size\n");
+          goto cleanup;
+        }
+      } else {
+        exponential_delay_factor *= 2;
+        log_warning(logger_id, "Local snapshot is delayed in %d ms, error code: %d\n",
+                    exponential_delay_factor * LOCAL_SNAPSHOTS_RESCAN_INTERVAL_MS, err);
+        cond_handle_timedwait(&lsm->cond_local_snapshots, &lock_cond,
+                              exponential_delay_factor * LOCAL_SNAPSHOTS_RESCAN_INTERVAL_MS);
+      }
+    }
+    if (!iota_local_snapshots_manager_should_take_snapshot(lsm, &tangle)) {
+      cond_handle_timedwait(&lsm->cond_local_snapshots, &lock_cond, LOCAL_SNAPSHOTS_RESCAN_INTERVAL_MS);
+      skip_check = false;
+    } else {
+      skip_check = true;
+    }
+  }
+
+cleanup:
+
+  lock_handle_unlock(&lock_cond);
+  lock_handle_destroy(&lock_cond);
+  if (iota_tangle_destroy(&tangle) != RC_OK) {
+    log_critical(logger_id, "Destroying tangle connection failed\n");
+  }
+
+  if (err != RC_OK) {
+    log_warning(logger_id, "Local snapshot is failed with error code: %d\n", err);
+  }
+
+  return NULL;
+}
+
+bool iota_local_snapshots_manager_should_take_snapshot(local_snapshots_manager_t const *const lsm,
+                                                       tangle_t const *const tangle) {
+  size_t new_transactions_count;
+  if (iota_tangle_transaction_count(tangle, &new_transactions_count) != RC_OK) {
+    log_critical(logger_id, "Failed in querying db size\n");
+    return false;
+  }
+
+  uint64_t latest_to_initial_gap = lsm->snapshots_service->snapshots_provider->latest_snapshot.metadata.index -
+                                   lsm->snapshots_service->snapshots_provider->inital_snapshot.metadata.index;
+
+  if ((latest_to_initial_gap > SNAPSHOT_SERVICE_MAX_NUM_MILESTONES_TO_CALC) ||
+      (((new_transactions_count - lsm->last_snapshot_transactions_count) >=
+        lsm->conf->local_snapshots.transactions_growth_threshold) &&
+       (latest_to_initial_gap > 2 * lsm->conf->local_snapshots.min_depth))) {
```

Comment:
Good questions, 
Short answer: because we want next snapshot to be `min_dpeth` deep
Long answer: The new target snapshot index is always  latest - `min_depth`, that means that initial milestone's index is "latest  - min_depth", so that condition is always met, meaning we can have a snapshot too often, *2 is too allow latest take min_depth milestones forward, and then promote initial from "latest - 2*min_depth" to  "latest - min_depth"